src/runtime-rs/crates/service/src/task_service.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    convert::{TryFrom, TryInto},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use common::types::{Request, Response};
+use containerd_shim_protos::{api, shim_async};
+use ttrpc::{self, r#async::TtrpcContext};
+
+use runtimes::RuntimeHandlerManager;
+
+pub(crate) struct TaskService {
+    handler: Arc<RuntimeHandlerManager>,
+}
+
+impl TaskService {
+    pub(crate) fn new(handler: Arc<RuntimeHandlerManager>) -> Self {
+        Self { handler }
+    }
+}
+
+async fn handler_message<TtrpcReq, TtrpcResp>(
+    s: &RuntimeHandlerManager,
+    ctx: &TtrpcContext,
+    req: TtrpcReq,
+) -> ttrpc::Result<TtrpcResp>
+where
+    Request: TryFrom<TtrpcReq>,
+    <Request as TryFrom<TtrpcReq>>::Error: std::fmt::Debug,
+    TtrpcResp: TryFrom<Response>,
+    <TtrpcResp as TryFrom<Response>>::Error: std::fmt::Debug,
+{
+    let r = req
+        .try_into()
+        .map_err(|err| ttrpc::Error::Others(format!("failed to translate from shim {:?}", err)))?;
+    let logger = sl!().new(o!("steam id" =>  ctx.mh.stream_id));
+    debug!(logger, "====> task service {:?}", &r);
+    let resp = s
+        .handler_message(r)
+        .await
+        .map_err(|err| ttrpc::Error::Others(format!("failed to handler message {:?}", err)))?;
+    debug!(logger, "<==== task service {:?}", &resp);
+    Ok(resp
+        .try_into()
+        .map_err(|err| ttrpc::Error::Others(format!("failed to translate to shim {:?}", err)))?)
+}
+
+#[async_trait]
+impl shim_async::Task for TaskService {
+    async fn state(
+        &self,
+        ctx: &TtrpcContext,
+        req: api::StateRequest,
+    ) -> ttrpc::Result<api::StateResponse> {
+        handler_message(&self.handler, ctx, req).await
+    }
```

Comment:
These `Task` trait methods almost have the same pattern, maybe implied by a macro.
But for now, you can leave a TODO in comment and optimize it next time.

Reply:
Thanks for your review, very good suggestion, modified as suggested.